// typestyle/lib/csx/layer.js
"use strict";
/**
 * @module Layers essentially allow you to create a new surface for layouts
 */
var _1 = require("../");
/**
 * New layer parent
 */
exports.layerParent = {
    position: 'relative'
};
/**
 * Use this to attach to any parent layer
 * and then you can use `left`/`top` etc to position yourself
 */
exports.attachToLayerParent = {
    position: 'absolute'
};
/**
 * This new layer will attach itself to the nearest parent with `position:relative` or `position:absolute`
 * And will become the new `layerParent`
 */
exports.newLayer = _1.extend(exports.attachToLayerParent, {
    left: 0,
    right: 0,
    top: 0,
    bottom: 0
});
exports.attachToTop = _1.extend(exports.attachToLayerParent, {
    top: 0,
    left: 0,
    right: 0
});
exports.attachToRight = _1.extend(exports.attachToLayerParent, {
    top: 0,
    right: 0,
    bottom: 0
});
exports.attachToBottom = _1.extend(exports.attachToLayerParent, {
    right: 0,
    bottom: 0,
    left: 0
});
exports.attachToLeft = _1.extend(exports.attachToLayerParent, {
    top: 0,
    bottom: 0,
    left: 0
});
/**
 * Helps fixing to page
 */
var fixed = {
    position: 'fixed'
};
exports.pageTop = _1.extend(fixed, {
    top: 0,
    left: 0,
    right: 0
});
exports.pageRight = _1.extend(fixed, {
    top: 0,
    right: 0,
    bottom: 0
});
exports.pageBottom = _1.extend(fixed, {
    right: 0,
    bottom: 0,
    left: 0
});
exports.pageLeft = _1.extend(fixed, {
    top: 0,
    bottom: 0,
    left: 0
});
